lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/launch/JavaLaunchCommandBuilder.java

```diff
@@ -25,17 +25,30 @@
 import java.io.File;
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 public final class JavaLaunchCommandBuilder implements LaunchCommandBuilder {
+  private static final Logger LOG = Logger.getLogger(JavaLaunchCommandBuilder.class.getName());
+
   private static final String DEFAULT_JAVA_PATH = System.getenv("JAVA_HOME") + "/bin/" + "java";
   private String stderrPath = null;
   private String stdoutPath = null;
-  private int megaBytes = 0;
   private String evaluatorConfigurationPath = null;
   private String javaPath = null;
   private String classPath = null;
   private Boolean assertionsEnabled = null;
+  private Map<String, JVMOption> options = new HashMap<>();
+  private Map<String, JVMOption> defaultOptions = new HashMap<String, JVMOption>() {{
```

Comment:
@senderista @markusweimer those are valid questions. Let me try to answer.

I've tried to maintain backwards-compatibility with the JVMHeapSlack parameter given by the runtimes. The problem is that the runtimes don't pre-set, rather they set the memory *after* the user has a change to call `setMemory` (in `onNext(ResourceLaunchEvent)`). If the runtimes just use `setMemory` there, then it would overwrite the user's memory setting.

My thinking in exposing the "default"/"user" approach was that in general runtimes might want to set some baselines other than just memory.

Another approach I thought of was to expose whether an option such as "-Xmx" is set. Then the runtime could check and set. This approach runs into a practical problem because java options don't cleanly separate into key-value pairs. For example, it might be difficult to make obvious what the query string would be to find if "-XX:+PrintGCDetails" is set. Also it seems to make implementing new runtimes just a tad bit more complex.